xhistogram/core.py

```diff
@@ -222,54 +274,77 @@ def histogram(
                 ax_positive = ndim + ax
             assert ax_positive < ndim, "axis must be less than ndim"
             axis_normed.append(ax_positive)
-        axis = np.atleast_1d(axis_normed)
+        axis = [int(i) for i in axis_normed]
 
-    do_full_array = (axis is None) or (set(axis) == set(range(a0.ndim)))
-    if do_full_array:
-        kept_axes_shape = None
-    else:
-        kept_axes_shape = tuple([a0.shape[i] for i in range(a0.ndim) if i not in axis])
+    all_arrays = list(args)
+    n_inputs = len(all_arrays)
 
-    all_args = list(args)
     if weights is not None:
-        all_args += [weights]
-    all_args_broadcast = broadcast_arrays(*all_args)
+        all_arrays.append(weights)
+        has_weights = True
+    else:
+        has_weights = False
 
-    def reshape_input(a):
-        if do_full_array:
-            d = a.ravel()[None, :]
-        else:
-            # reshape the array to 2D
-            # axis 0: preserved axis after histogram
-            # axis 1: calculate histogram along this axis
-            new_pos = tuple(range(-len(axis), 0))
-            c = np.moveaxis(a, axis, new_pos)
-            split_idx = c.ndim - len(axis)
-            dims_0 = c.shape[:split_idx]
-            assert dims_0 == kept_axes_shape
-            dims_1 = c.shape[split_idx:]
-            new_dim_0 = np.prod(dims_0)
-            new_dim_1 = np.prod(dims_1)
-            d = reshape(c, (new_dim_0, new_dim_1))
-        return d
+    dtype = "i8" if not has_weights else weights.dtype
 
-    all_args_reshaped = [reshape_input(a) for a in all_args_broadcast]
+    bins = _ensure_bins_is_a_list_of_arrays(bins, n_inputs)
 
-    if weights is not None:
-        weights_reshaped = all_args_reshaped.pop()
-    else:
-        weights_reshaped = None
+    bincount_kwargs = dict(weights=has_weights, axis=axis, bins=bins, density=density)
 
-    n_inputs = len(all_args_reshaped)
-    bins = _ensure_bins_is_a_list_of_arrays(bins, n_inputs)
+    # here I am assuming all the arrays have the same shape
+    # probably needs to be generalized
+    input_indexes = [tuple(range(a.ndim)) for a in all_arrays]
+    input_index = input_indexes[0]
+    assert all([ii == input_index for ii in input_indexes])
```

Comment:
nit: it seems this is only needed in the `if _any_dask_array` case; maybe move it into that conditional?

> here I am assuming all the arrays have the same shape

Maybe you'd want to `np.broadcast_arrays(*all_arrays)` here (note that this will dispatch to `da.broadcast_arrays` automatically thanks to NEP-18)?

Alternatively, I think you could get away without broadcasting with some slight modifications to the blockwise indices. Specifically, rather than using `input_index` for all arrays, use each array's own indices. Then blockwise will basically express any the broadcasting for you in the graph:

```python
input_indexes = [tuple(range(a.ndim)) for a in all_arrays]
largest_index = max(input_indexes, key=len)
if axis is not None:
    drop_axes = tuple([ii for ii in largest_index if ii in axis])
else:
    drop_axes = largest_index

out_index = largest_index + tuple(new_axes)
...
blockwise_args = []
for arg, index in zip(all_arrays, input_indexes):
    blockwise_args.append(arg)
    blockwise_args.append(index)
```

Reply:
@gjoseph92 is there any performance reason to prefer making blockwise handle the broadcasting? It seems simplest to just broadcast everything at the start, but does that mean carrying around larger arrays than necessary?

Reply:
The arrays shouldn't actually increase in physical size; their shape will just be expanded. The only advantage of using blockwise alone is that the graph will be a little smaller:
* No extra reshape layers added prior to the blockwise (which are also fully materialized layers and slightly more expensive to serialize to the scheduler)
* The low-level graphs of the arrays input into blockwise won't contain superfluous keys for the extra broadcast dimensions

In practice, I very much doubt it would be an appreciable difference, but not having compared them I can't say for sure. There is also a slight optimization advantage with pure-blockwise when there is exactly 1 input array, and it's already the result of a blockwise operation.

Reply:
That's interesting, thanks.

However it's just occurred to me that surely relying on blockwise to do all the broadcasting is not an option, simply because the non-dask case doesn't use blockwise. Supporting a numpy-only code path (as we would likely want both for debugging and for fitting xarray's standard practice) means broadcasting all arrays before they are fed to `_bincount`.